### DISCUSSION TITLE
How to use extra keys on Corne v4

### DISCUSSION THREAD

**Discussion by bogorad** [manna-harbour/miryoku#373] [Q&A] [answered]
I installed qmk, then [Drashna's qmk_userpace](https://github.com/drashna/qmk_userspace/tree/miryoku) (miryoku branch). Compiles ok, and everything seems to work.

Question: how do I use the four extra keys on Corne V4? These keys are not present in 

https://github.com/foostan/kbd_firmware/blob/main/keyboards/crkbd/qmk/qmk_firmware/keymaps/default/keymap.c

only in ../via/ 

https://github.com/foostan/kbd_firmware/blob/main/keyboards/crkbd/qmk/qmk_firmware/keymaps/via/keymap.c

where they are defined as `LAYOUT_split_3x6_3_ex2`

What do I need to do to be able to use these keys in Miryoku?

**Comment by manna-harbour** [comment, marked as answer]
For QMK, they are in https://github.com/qmk/qmk_firmware/blob/master/keyboards/crkbd/rev4_0/info.json.

To use with Miryoku QMK, this requires that a mapping is added below https://github.com/qmk/qmk_firmware/tree/master/keyboards/crkbd. For example see 
https://github.com/manna-harbour/miryoku_qmk/blob/miryoku/layouts/community/split_3x6_3/manna-harbour_miryoku/config.h. To make that easier to produce you'd first use https://docs.qmk.fm/newbs_building_firmware#create-a-new-keymap.

Normally this can also be added by request, but for QMK that would need to wait for https://github.com/manna-harbour/miryoku/discussions/287 to be dealt with first.

**Reply by manna-harbour** [reply]
> To make that easier to produce you'd first use https://docs.qmk.fm/newbs_building_firmware#create-a-new-keymap.

Actually, that doesn't help at all, as `qmk new-keymap` doesn't support `FORCE_LAYOUT`, and editing `info.json` doesn't work either as it ?just copies the default keymap anyway...

So, it seems the QMK way is to download a json keymap for the layout from the configurator https://config.qmk.fm/#/crkbd/rev4_0/standard/LAYOUT_split_3x6_3_ex2, and convert with `qmk json2c`.

That will give a keymap like this:
```
    [0] = LAYOUT_split_3x6_3_ex2(\
KC_NO, KC_Q, KC_W, KC_E, KC_R, KC_T, KC_F1,     KC_F3, KC_Y, KC_U, KC_I, KC_O, KC_P, KC_NO, \
KC_NO, KC_A, KC_S, KC_D, KC_F, KC_G, KC_F2,     KC_F4, KC_H, KC_J, KC_K, KC_L, KC_SCLN, KC_NO, \
KC_NO, KC_Z, KC_X, KC_C, KC_V, KC_B,                   KC_N, KC_M, KC_COMM, KC_DOT, KC_SLSH, KC_NO, \
                KC_NO, KC_NO, KC_NO,                   KC_NO, KC_NO, KC_NO \
)
};
```
So, the extra keys are added to the top and middle rows. The `split_3x6_3` mapping can then be adapted like this:

```
// Copyright 2025 Manna Harbour
// https://github.com/manna-harbour/miryoku

// This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 2 of the License, or (at your option) any later version. This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with this program. If not, see <http://www.gnu.org/licenses/>.

#pragma once

#define XXX KC_NO

#define LAYOUT_miryoku( \
      K00,  K01,  K02,  K03,  K04,                     K05,  K06,  K07,  K08,  K09, \
      K10,  K11,  K12,  K13,  K14,                     K15,  K16,  K17,  K18,  K19, \
      K20,  K21,  K22,  K23,  K24,                     K25,  K26,  K27,  K28,  K29, \
      N30,  N31,  K32,  K33,  K34,                     K35,  K36,  K37,  N38,  N39 \
) \
LAYOUT_split_3x6_3_ex2( \
XXX,  K00,  K01,  K02,  K03,  K04,  XXX,         XXX,  K05,  K06,  K07,  K08,  K09,  XXX, \
XXX,  K10,  K11,  K12,  K13,  K14,  XXX,         XXX,  K15,  K16,  K17,  K18,  K19,  XXX, \
XXX,  K20,  K21,  K22,  K23,  K24,                     K25,  K26,  K27,  K28,  K29,  XXX , \
                  K32,  K33,  K34,                     K35,  K36,  K37 \
)
```

After adding the mapping you can the populate the extra keys with https://github.com/manna-harbour/miryoku/discussions/85. You'll also need to use `FORCE_LAYOUT` when building.

**Reply by Zev18** [reply]
Where does this `LAYOUT_split_3x6_3_ex2` go? Does the corne `info.json` need to be modified as well?

**Reply by manna-harbour** [reply]
@Zev18 Linking to the other discussion: https://github.com/manna-harbour/miryoku/discussions/380.
